### PR TITLE
CR-1028065 qdma: add per stream aio context for polling individual stream

### PR DESCRIPTION
**Pull request by karenx-xilinx**
Added a new API to poll individual stream: clPollStream(), and internally, added a per-queue aio context for io submission and completion polling
Tested with u200 image from 2020.1_qualified_latest.

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 1]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+
+    memset(&qAioCtx, 0, sizeof(qAioCtx));
+    qAioEn = (io_setup(SHIM_QDMA_AIO_EVT_MAX, &qAioCtx) == 0);
+}
+
+int queue_cb::queue_poll_completion(unsigned int min_compl,
+                unsigned int max_compl, struct xclReqCompletion *comps,
+                unsigned int* actual, unsigned int timeout /*ms*/)
+{
+    struct timespec time, *ptime = NULL;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: nullptr (don't use NULL in C++ code)

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 2]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+    unsigned int num_evt, i;
+    int rc;
+
+    *actual = 0;
+    if (!qAioEn) {
+        num_evt = -EINVAL;
+        std::cerr << "async io is NOT enabled" << std::endl;
+        goto done;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: Lets avoid 'goto',  just return here and no need for num_evt.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 3]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+    }
+
+    rc = io_getevents(qAioCtx, min_compl, max_compl, (struct io_event *)comps, ptime);
+    if (rc < 0)
+        return rc;
+
+    num_evt = rc;
+    if (num_evt < min_compl) {
+        std::cerr << "failed to poll queue" << std::endl;
+        goto done;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: How about just returning 'rc'?

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 4]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+    }
+    *actual = num_evt;
+    for (i = num_evt - 1; i >= 0; i--) {
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: for (int i = ...)

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 5]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+            comps[i].nbytes = 0;
+            comps[i].err_code = ((struct io_event *)comps)[i].res;
+        } else {
+            comps[i].nbytes = ((struct io_event *)comps)[i].res;
+            comps[i].err_code = ((struct io_event *)comps)[i].res2;
+        }
+    }
+    num_evt = 0;
+
+done:
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: Remove

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 6]
```diff
@@ -123,6 +123,140 @@ inline int io_getevents(aio_context_t ctx, long min_nr, long max_nr,
...
+                unsigned int* actual, unsigned int timeout /*ms*/)
+{
+    struct timespec time, *ptime = NULL;
+    unsigned int num_evt, i;
+    int rc;
+
+    *actual = 0;
+    if (!qAioEn) {
+        num_evt = -EINVAL;
+        std::cerr << "async io is NOT enabled" << std::endl;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: use xrt_logmsg for all errors, see other places in this file.

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 7]
```diff
@@ -1226,10 +1360,12 @@ int shim::xclCreateWriteQueue(xclQueueContext *q_ctx, uint64_t *q_hdl)
     rc = ioctl(mStreamHandle, XOCL_QDMA_IOC_CREATE_QUEUE, &q_info);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: int rc = ...

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 8]
```diff
@@ -1249,23 +1385,28 @@ int shim::xclCreateReadQueue(xclQueueContext *q_ctx, uint64_t *q_hdl)
...
+    queue_cb *qcb = reinterpret_cast<queue_cb *>(q_hdl);
     int rc;
 
-    rc = close((int)q_hdl);
+    rc = close((int)qcb->queue_get_hndl());
     if (rc)
         xrt_logmsg(XRT_ERROR, "%s: Destroy Queue failed", __func__);
 
+
+    free(qcb);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: Please use delete to match new

**Comment by karenx-xilinx** [comment]
Thanks, Soren. I will submit a new patch with the changes.

**Pull request closed without merging**